File: app/reader.js
// Bounds-checked buffer reads — the JS mirror of the Python defensive helpers.
// Every parse read goes through these: an out-of-range access returns null (or "")
// instead of throwing or reading past the buffer, so a malformed/short save
// degrades to "unknown" rather than crashing. Buffers are Uint8Array throughout.

/** Little-endian unsigned int, or null if the read runs past the end. */
export function readUint(buf, off, size) {
  if (off == null || off < 0 || off + size > buf.length) return null;
  let v = 0;
  for (let i = size - 1; i >= 0; i--) v = v * 256 + buf[off + i];
  return v;
}

export const u8 = (buf, off) => readUint(buf, off, 1);
export const u16 = (buf, off) => readUint(buf, off, 2);
export const u32 = (buf, off) => readUint(buf, off, 4);
export const u64 = (buf, off) => readUint(buf, off, 8);

const UTF16 = new TextDecoder("utf-16le", { fatal: false });

/**
 * Decode a UTF-16LE string ending at the first null pair. Mirrors the Python
 * `read_utf16`, including its quirk: the null pair is found at ANY byte index
 * (not only even), then the slice is kept byte-pair aligned via `end + (end & 1)`.
 */
export function readUtf16(buf, off, maxChar) {
  if (off == null || off < 0 || off >= buf.length) return "";
  const raw = buf.subarray(off, Math.min(off + maxChar * 2, buf.length));
  let end = -1;
  for (let i = 0; i + 1 < raw.length; i++) {
    if (raw[i] === 0 && raw[i + 1] === 0) { end = i; break; }
  }
  const slice = end !== -1 ? raw.subarray(0, end + (end & 1)) : raw;
  return UTF16.decode(slice).replace(/\u0000+$/, "");
}

const NAME_OK = new Set(
  "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 -_'".split("")
);

/** Is a decoded string a plausible character name (non-empty, allowed chars only)? */
export function isValidName(name) {
  if (!name) return false;
  for (const c of name) if (!NAME_OK.has(c)) return false;
  return true;
}

/** First index of `pattern` (Uint8Array) in `buf` at or after `start`, or -1. */
export function indexOf(buf, pattern, start = 0) {
  const n = buf.length, m = pattern.length;
  if (m === 0) return start;
  for (let i = Math.max(0, start); i + m <= n; i++) {
    let ok = true;
    for (let j = 0; j < m; j++) {
      if (buf[i + j] !== pattern[j]) { ok = false; break; }
    }
    if (ok) return i;
  }
  return -1;
}

File: sl2/reader.py
"""Safe, bounds-checked readers.

Nothing anywhere in the package indexes a buffer without going through these. A
read that would run off the end returns None (or "") instead of raising or
reading whatever happens to sit past it.
"""


##
# @brief Read a little-endian unsigned integer, or None if it would run past the
#        end of the buffer.
# @param buf  The bytes to read from.
# @param off  Byte offset. A negative offset is treated as out of range.
# @param size Width in bytes (1, 2, 4, or 8).
# @return The integer value, or None if the read is out of range.
def read_uint(buf, off, size):
    if off is None or off < 0 or off + size > len(buf):
        return None
    return int.from_bytes(buf[off:off + size], "little")


## @brief One-byte read. @see read_uint
def u8(buf, off):
    return read_uint(buf, off, 1)


## @brief Two-byte read. @see read_uint
def u16(buf, off):
    return read_uint(buf, off, 2)


## @brief Four-byte read. @see read_uint
def u32(buf, off):
    return read_uint(buf, off, 4)


## @brief Eight-byte read. @see read_uint
def u64(buf, off):
    return read_uint(buf, off, 8)


##
# @brief Decode a UTF-16LE string that ends at the first null pair.
# @details Souls names are UTF-16LE and not always fixed-length, so this reads a
# bounded window and stops at the first @c 0x0000. Returns an empty string on a
# bad read rather than raising.
# @param buf      The bytes to read from.
# @param off      Where the string starts.
# @param max_char Maximum characters to consider.
# @return The decoded string, stripped of trailing nulls.
def read_utf16(buf, off, max_char):
    if off is None or off < 0 or off >= len(buf):
        return ""
    raw = buf[off:off + max_char * 2]
    end = raw.find(b"\x00\x00")
    if end != -1:
        raw = raw[:end + (end & 1)]  # keep byte pairs aligned
    try:
        return raw.decode("utf-16-le", "ignore").rstrip("\x00")
    except (UnicodeDecodeError, ValueError):
        return ""


## @brief The only characters a real player name may contain. Anything outside
#         this set means the bytes are not a name — usually an empty slot.
NAME_OK = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 -_'")


##
# @brief Decide whether a decoded string is a plausible character name.
# @param name The candidate string.
# @return True if it is non-empty and every character is allowed.
def is_valid_name(name):
    return bool(name) and all(c in NAME_OK for c in name)
